# Entropy

In [1]:
# import libraries
import numpy as np
import matplotlib.pyplot as plt

## Reminder of entropy

$$H(p) = -\sum_x p(x)\log(p(x))$$

In [3]:
# Probabilty of an event happening 
p = .25

# WRONG FORMULA!
H = - (p*np.log(p))
print('Wrong entropy: ' + str(H))

Wrong entropy: 0.34657359027997264


The formula above is wrong because we calculated **the chance of the event happeing** but NOT **the chance of the event not happening**

In [4]:
# corrected formula
x = [.25, .75]

H = 0
for p in x:
    H -= p * np.log(p)

print('Correct entropy: ' + str(H))

Correct entropy: 0.5623351446188083


In [5]:
# also correct, written out for N=2 events
H = -( p*np.log(p) + (1-p)*np.log(1-p) )
print('Correct entropy: ' + str(H))

Correct entropy: 0.5623351446188083


# Cross-Entropy

In [15]:
# probabilities MUST sum to 1.0
p = [1, 0] 
q = [.25 , .75]

H = 0
for i in range(len(p)):
    H -= p[i] * np.log(q[i])

print('Cross entropy: ' + str(H))

Cross entropy: 1.3862943611198906


In [11]:
# loopless form
# also correct, written out for N=2 events
H = -( p[0]*np.log(q[0]) + p[1]*np.log(q[1]) )
print('Correct entropy: ' + str(H))

Correct entropy: 1.3862943611198906


Notice how `p = [1, 0]`. the second term is `0`. which means the term will be cancelled. we can simplify it. also `1` can be removed because `n * 1 = n`.

In [10]:
# simplification
H = -np.log(q[0])
print('Manually simplified: ' + str(H))

Manually simplified: 1.3862943611198906


The formula we got now is considered **Binary Cross-Entropy**.

## Cross-entropy using pytorch

In [17]:
import torch
import torch.nn.functional as F

# convert lists to tensors
p_tensor = torch.Tensor(p)
q_tensor = torch.Tensor(q)

# apply bce
F.binary_cross_entropy(p_tensor,q_tensor)

tensor(75.)

notice how the answer is different. its not `1.386...`. That's because pytorch made this function inverted.

It expects `(model_predicted_probability, label)`

since p is our labels and q is prediction it came out wrong. So let's correct it.

In [18]:
F.binary_cross_entropy(q_tensor,p_tensor)

tensor(1.3863)